Support MySQL 8 and fix DB, refs #13220 and #13221

jraddaoui opened the pull request:
**Update Docker environment:**

- Use `percona:8.0` image.
- Use `mysql_native_password` as `default-authentication-plugin`.
- Change default charset to `utf8mb4` and collation to `utf8mb4_0900_ai_ci`.
- Remove `NO_AUTO_CREATE_USER` from `sql_mode`.

**Rename QubitFunction model:**

In MySQL 8, `function` became a reserved word causing errors when
working with QubitFunction due to it's auto-generated table name.

- Change schema declaration to `function_object`.
- Generate SQL and models and remove old ones.
- Change references to old models.
- Update ES mapping and code that depends on the model class name.
- Add migration to change existing database tables and class reference.
- Avoid changing module name and related settings to keep the front-end
  representation (URLs and labels) as they were.
- Change model class reference in old migration classes/process. This
  process should be executed with the 1.1 code to upgrade to 1.1 before
  running the actual upgrade task. Therefore, this change is only
  cosmetic, to avoid finding references to the old model class.

**Normalize database schema:**

- Avoid differences between upgraded and new install database schema.
- Do it before the `utf8mb4` migration to fix NULL columns.
- Improve error management on migrations:
  - Split statements to show errors on arMigration0169.
  - Throw exception when expected indexes or foreign keys are missing
    on arMigration0176 and arMigration0177.

**Migrate to `utf8mb4`:**

The utf8mb3 character set is deprecated and will be removed in a future
MySQL release.

- Change default database and connection encoding.
- Use `utf8mb4_0900_ai_ci` collation.
- Reduce DO path index size in chars to allow the extra byte.
- Convert existing data using MODIFY over char columns.
- Analyze all tables after conversion.

jraddaoui commented:
About the `function` table name conflict:

Ideally, we would modify the schema and create a migration to rename the table as we wouldn't have use that name knowing it's a reserved word. However, there are several queries in `QubitMigrate` in which the `function` table name may be passed as a variable and those queries can be executed before the new migration is reached. Therefore, we should revisit all raw queries and add backquotes in those where the `function` table is used or can be passed as a variable (PARTIALLY DONE). Additionally, we must remove the use of the `function` word as an alias (DONE?).

After adding the backquotes, changing the table name shouldn't be necessary and I'm having conflicts about doing it or not. On one hand, it can avoid problems and confusions in the future but, on the other, it may not be an easy task as we'll need to modify the schema and model and that could also cause issues on the existing migrations.

djjuhasz commented on docker/etc/mysql/mysqld.cnf:
```diff
@@ -1,7 +1,8 @@
-character-set-server=utf8
-collation-server=utf8_unicode_ci
+character-set-server=utf8mb4
+collation-server=utf8mb4_unicode_ci
```
Shouldn't this be `collation-server=utf8mb4_0900_ai_ci`?

jraddaoui replied:
Good point @djjuhasz!

I didn't want to change the current collation that much but it may be a good idea to use the latest Unicode version and the case and accent insensitivity.

Extra reading:
https://dev.mysql.com/doc/refman/8.0/en/charset-unicode-sets.html
https://mysqlserverteam.com/new-collations-in-mysql-8-0-0/
https://stackoverflow.com/questions/54885178/whats-the-difference-between-utf8-unicode-ci-and-utf8mb4-0900-ai-ci

djjuhasz commented:
I'm in favor of sticking with "function" for the table name, unless a good alternative is proposed. I looked at the definition of function in ISDF (quoted below) and I can't see any immediately apparent alternative to "function".

> **Function**.    Any  high  level  purpose,  responsibility  or  task  assigned  to  the  accountability agenda of a corporate body by legislation, policy or mandate. Functions may be decomposed into sets of co-ordinated operations such as subfunctions, business processes, activities, tasks or transactions.

-- https://www.ica.org/sites/default/files/CBPS_2007_Guidelines_ISDF_First-edition_EN.pdf, retrieved 2019-11-06

djjuhasz commented:
@jraddaoui thanks for providing all of the helpful reference links!  Lots of interesting information and considerations in upgrading. :)

mcantelon commented:
Yeah, looks like there's about 17 instances of "function." being used to reference the table so seems updateable.

jraddaoui commented:
> I'm in favor of sticking with "function" for the table name, unless a good alternative is proposed. I looked at the definition of function in ISDF (quoted below) and I can't see any immediately apparent alternative to "function".
> 
> > **Function**.    Any  high  level  purpose,  responsibility  or  task  assigned  to  the  accountability agenda of a corporate body by legislation, policy or mandate. Functions may be decomposed into sets of co-ordinated operations such as subfunctions, business processes, activities, tasks or transactions.
> 
> -- https://www.ica.org/sites/default/files/CBPS_2007_Guidelines_ISDF_First-edition_EN.pdf, retrieved 2019-11-06

We wouldn't have to change the visual name, just the model on the backend (`QubitFunctionObject` for example). But I think it's better if we can avoid changing it, and having backquotes in the raw queries table names doesn't hurt and may help us in the future.

jraddaoui commented:
> Yeah, looks like there's about 17 instances of "function." being used to reference the table so seems updateable.

Thanks @mcantelon! Could you share where are they in the code? Or what did you use to search those occurrences?

mcantelon commented:
I just used a gigantic grep statement, filtering out stuff in which a reference to "function." didn't seem to be for SQL:

`grep -r "function\." * | grep -v "\.log" | grep -v "\.js" | grep -v "messages\.xml" | grep -v "\.pkg" | grep -v "MailTransport.php" | grep -v "HTMLsymbol.ent" | grep -v "MailInvoker.php" | grep -v "xhtml-symbol.ent" | grep -v "saxon9he.jar" | grep -v "FunctionScore.php" | grep -v "JavascriptHelper.php" | grep -v "sfViewCacheManager.class.php" | grep -v "sfAutoloadAgain.class.php" | grep -v "HTMLModule.php" | grep -v "EasyRdf" | grep -v "htmlpurifier" | grep -v "sfTestFunctionalBase.class.php" | grep -v "Gearman" | grep -v "Doctrine" | grep -v "QubitMarkdown.class.php" | grep -v "DBAdapter.php" | grep -v "sfPropelPlugin" | grep -v "sfSkosUniqueRelations.class.php" | grep -v "sfCultureInfo.class.php" | grep -v "miniColors" | grep -v "swiftmailer" | grep -v "FluentDOM" | grep -v "dtd" | grep -v "Qubit\.class\.php" | grep -v "QubitDigitalObject\.php" | grep -v "QubitMigrate\.class\.php" | grep -v "QubitXmlImport\.class\.php" | grep -v "_relatedFunction\.php" | grep -v "editSuccess\.php" | wc -l`

jraddaoui commented:
Hi @djjuhasz and @mcantelon,

This is finally ready for code review. I had some back and forwards along the process so I'd suggest that you read the updated PR description and jump into the full code changes instead of going commit by commit.

Also, https://projects.artefactual.com/issues/13221 has more information about some schema differences I found during the upgrade that I addressed in this PR too.

djjuhasz commented on lib/task/migrate/migrations/arMigration0177.class.php:
```diff
@@ -0,0 +1,245 @@
+      $sql = 'UPDATE %s SET culture=:cul WHERE culture IS NULL;';
+      QubitPdo::modify(
+        sprintf($sql, $i18nTable),
+        array(':cul' => $defaultCulture)
```
I would suggest setting the `source_culture` first (e.g. if source_culture is NULL, set it to default culture) *then* setting the `i18n.culture = source_culture`.   In most cases if there is no culture set, I would assume the culture should be the same as the row source_culture, rather than the default culture.  

Practically, I'm sure in most cases when the culture == null the source_culture == null too but I'm not sure if we can assume it.

jraddaoui replied:
Good catch, thanks @djjuhasz!

djjuhasz reviewed (commented):
Looks great @jraddaoui! 😄   I made one suggestion, but it may or may not actually make a difference to the upgrade.

djjuhasz reviewed (approved):
Looks great! :shipit:

jraddaoui committed:
Support MySQL 8 and fix DB, refs #13220 and #13221

Update Docker environment:

- Use `percona:8.0` image.
- Use `mysql_native_password` as `default-authentication-plugin`.
- Change default charset to `utf8mb4` and collation to `utf8mb4_0900_ai_ci`.
- Remove `NO_AUTO_CREATE_USER` from `sql_mode`.

Rename QubitFunction model:

In MySQL 8, `function` became a reserved word causing errors when
working with QubitFunction due to it's auto-generated table name.

- Change schema declaration to `function_object`.
- Generate SQL and models and remove old ones.
- Change references to old models.
- Update ES mapping and code that depends on the model class name.
- Add migration to change existing database tables and class reference.
- Avoid changing module name and related settings to keep the front-end
  representation (URLs and labels) as they were.
- Change model class reference in old migration classes/process. This
  process should be executed with the 1.1 code to upgrade to 1.1 before
  running the actual upgrade task. Therefore, this change is only
  cosmetic, to avoid finding references to the old model class.

Normalize database schema:

- Avoid differences between upgraded and new install database schema.
- Do it before the `utf8mb4` migration to fix NULL columns.
- Improve error management on migrations:
  - Split statements to show errors on arMigration0169.
  - Throw exception when expected indexes or foreign keys are missing
    on arMigration0176 and arMigration0177.

Migrate to `utf8mb4`:

The utf8mb3 character set is deprecated and will be removed in a future
MySQL release.

- Change default database and connection encoding.
- Use `utf8mb4_0900_ai_ci` collation.
- Reduce DO path index size in chars to allow the extra byte.
- Convert existing data using MODIFY over char columns.
- Analyze all tables after conversion.

Squashed commit of the following:

commit d6d56fbc0835029863244365d1ab6af9d43a8895
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Wed Dec 11 13:52:44 2019 +0100

    Fix source culture first

    And use the fixed value to update the missing i18n culture values.

commit 16a9d586bbc021280987744f265e8df04b704829
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Dec 6 16:57:57 2019 +0100

    Analyze tables after `utf8mb4` conversion

commit c4e40cb2a8e39c831553df3d9fe61fbd636e91e3
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Dec 6 16:57:15 2019 +0100

    Reformat migration 178

    Declare alter table statements in array to loop over it and other
    cosmetic changes.

commit 6cadaa413cec139a7abd43a52073bf387545ac9e
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Dec 6 14:39:27 2019 +0100

    Add `sql_mode` comment

commit e63573469eb421114c2ed127b5accef387d64d0a
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Fri Dec 6 14:06:13 2019 +0100

    Remove comment

    The NOT NULL restorations have been moved to the previous migration. In
    this, they're only maintained.

commit 6d744cbb5a88f1d6dcc2bfa4fdd24fb83b98bcff
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Thu Dec 5 23:06:06 2019 +0100

    Improve error management on migrations

    - Split statements to show errors on arMigration0169.
    - Throw exception when expected indexes or foreign keys are missing on
      arMigration0176 and arMigration0177.

commit 578adb3f49915eb3e254c96b288ce77ebe5d9849
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Thu Dec 5 18:32:32 2019 +0100

    Normalize database schema

    - Avoid differences between upgraded and new install database schema.
    - Do it before the `utf8mb4` migration to fix NULL columns.

commit 03d5a0ef55b088a15a35d1670c19256b84227f49
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Thu Dec 5 16:06:16 2019 +0100

    Use MODIFY to transform to utf8mb4

commit af6bea4b7ad1b94c22c80f2d9fde6dd55b710c22
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon Dec 2 11:52:29 2019 +0100

    Migrate to `utf8mb4`

    The utf8mb3 character set is deprecated and will be removed in a future
    MySQL release.

    - Change default database and connection encoding.
    - Use `utf8mb4_0900_ai_ci` collation.
    - Reduce DO path index size in chars to allow the extra byte.
    - Convert existing data to `utf8_mb4`.
    - TODO: check and update VARCHAR and TEXT type columns length.

commit 5ae9bfac53d7dbd9d5f04fc6dcd968883609794c
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon Dec 2 16:36:16 2019 +0100

    Remove back quote escaping fix

    After changing the `function` table name this fix is no longer needed.

commit 3f1443b03a9072ea992607249bc21f942f43efe4
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Mon Dec 2 16:27:40 2019 +0100

    Rename QubitFunction model

    In MySQL 8, `function` became a reserved word causing errors when
    working with QubitFunctions due to it's auto-generated table name.

    - Change schema declaration to `function_object`.
    - Generate SQL and models and remove old ones.
    - Change references to old models.
    - Update ES mapping and code that depends on the model class name.
    - Add migration to change existing database tables and class reference.
    - Avoid changing module name and related settings to keep the front-end
      representation (URLs and labels) as they were.
    - Change model class reference in old migration classes/process. This
      process should be executed with the 1.1 code to upgrade to 1.1 before
      running the actual upgrade task. Therefore, this change is only
      cosmetic, to avoid finding references to the old model class.

commit b6f724b08c0e74b5948c73846ab7277a5800cdbf
Author: José Raddaoui Marín <raddaouimarin@gmail.com>
Date:   Tue Nov 5 22:16:15 2019 +0100

    Support MySQL 8

    - Use `mysql_native_password` as `default-authentication-plugin`.
    - Change default charset and collation to `utf8mb4`.
    - Remove `NO_AUTO_CREATE_USER` from `sql_mode`.
    - The `function` table conflicts with a new reserved word.
